### PR TITLE
Fix Sleep/Purity checks being skipped under certain overrides

### PR DESCRIPTION
**Pull request by fira**
Long overdue PR for #267 and as discuted on coderbus ages ago back then

TL;DR is, there being an override one way or another shouldn't prevent you from checking children too

Discovered on CM-SS13 as we had a controller held up by sleep randomly in production, and no amount of hammering SHOULD_NOT_SLEEP checks seemed to trip anything !

As additional and separate note, and as we discussed back then on coderbus, there are more situations with assumptions of types that mean SpacemanDMM will miss out on a big part of the violators. It's actually for best, because "properly" checking is near impossible on SS13 due to double dispatch constructs just about erasing the "actual" types. This resulted in about 25,000 violating procs when fixing it for CM-SS13, for example...

### CODE REVIEW

**Comment by spookydonut** [comment]
@fira I'd like to see a test case written for this that proves it works and will fail if your changes are removed.

**Comment by fira** [comment]
> @fira I'd like to see a test case written for this that proves it works and will fail if your changes are removed.

there is a snippet in the issue but let me add a test proper then too
